### PR TITLE
[WIP] Kubeadm bootstrapper

### PR DESCRIPTION
**Pull request by r2d4**
Some TODOs

- [ ] Make custom storage provisioner work with the kubeadm bootstrapper.  Currently it doesn't since we run it directly in localkube.  I've skipped that particular integration test right now.
- [ ] Implement the "extra config" with the kubeadm config file
- [ ] Use an upstream version of kubeadm once one is released with the `phase alpha controlplane` command.  This was recently merged.
- [ ] Status should check the status of the apiserver.  Currently it just checks the status of the kubelet
- [ ] Logs should return logs of all the components of the control plane.  

TODOs probably in scope for this PR

- [ ] Add unit tests
- [x] Handle the none driver.  There are a few obvious fixes needed, but I haven't done any extensive testing.
- [ ] Figure out a persisted place for the etcd data directory (I just stuck it in `/data`)

### CODE REVIEW

File: cmd/minikube/cmd/config/config.go [hunk 1]
```diff
@@ -114,6 +116,10 @@ var settings = []Setting{
 		set:  SetString,
 	},
 	{
+		name: Bootstrapper,
+		set:  SetString, //TODO(r2d4): more validation here?
```
Comment (aaron-prindle) [inline cmd/minikube/cmd/config/config.go]: Perhaps validation function here that checks against a list of available bootstrappers

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+		} else if err != nil {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+func (k *KubeadmBootstrapper) RestartCluster(k8s bootstrapper.KubernetesConfig) error {
+	if err := k.c.Run("rm -rf /etc/kubernets/*.conf"); err != nil {
```
Comment (nkubala) [inline pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: should this be `/etc/kubernetes/*.conf`?

File: pkg/minikube/bootstrapper/localkube/localkube_test.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+	tests := []struct {
+		description string
+		follow      bool
+	}{
+		{
+			description: "logs -f",
+			follow:      true,
+		},
+		// TODO(r2d4): fix this test
```
Comment (r2d4) [inline pkg/minikube/bootstrapper/localkube/localkube_test.go]: Fix

File: pkg/minikube/constants/constants.go [hunk 4]
```diff
@@ -135,11 +136,31 @@ const (
 )
 
 const (
+	KubeletServiceFile     = "/lib/systemd/system/kubelet.service"
+	KubeletSystemdConfFile = "/etc/systemd/system/kubelet.service.d/10-kubeadm.conf"
+	KubeadmConfigFile      = "/var/lib/kubeadm.yaml"
```
Comment (r2d4) [inline pkg/minikube/constants/constants.go]: put in /etc/kubernetes

File: pkg/minikube/sshutil/sshutil.go [hunk 5]
```diff
@@ -166,10 +76,39 @@ func newSSHHost(d drivers.Driver) (*sshHost, error) {
 	}, nil
 }
 
-func DeleteFile(f assets.CopyableFile, client *ssh.Client) error {
-	return RunCommand(client, GetDeleteFileCommand(f))
-}
+func GetShell(session *ssh.Session, cmd string) error {
```
Comment (r2d4) [inline pkg/minikube/sshutil/sshutil.go]: Do we actually need this

File: pkg/minikube/sshutil/sshutil.go [hunk 6]
```diff
@@ -17,29 +17,17 @@ limitations under the License.
...
 	"github.com/docker/machine/libmachine/drivers"
 	machinessh "github.com/docker/machine/libmachine/ssh"
 	"github.com/pkg/errors"
 	"golang.org/x/crypto/ssh"
-	"k8s.io/minikube/pkg/minikube/assets"
-	"k8s.io/minikube/pkg/util"
 )
 
-// SSHSession provides methods for running commands on a host.
-type SSHSession interface {
```
Comment (r2d4) [inline pkg/minikube/sshutil/sshutil.go]: Where was this used?

File: pkg/minikube/sshutil/sshutil_test.go [hunk 7]
```diff
@@ -17,42 +17,43 @@ limitations under the License.
...
-		t.Fatalf("Error!")
-	}
-
-	if _, ok := s.Commands[cmd]; !ok {
-		t.Fatalf("Expected command: %s", cmd)
-	}
-}
+//TODO(r2d4): Move this test to bootstrapper package
+
+// func TestNewSSHClient(t *testing.T) {
```
Comment (r2d4) [inline pkg/minikube/sshutil/sshutil_test.go]: Fix this

File: pkg/minikube/bootstrapper/kubeadm/kubeadm.go (hunk 2 above)
Comment (r2d4) [inline pkg/minikube/bootstrapper/kubeadm/kubeadm.go]: Good catch.  Maybe I can remove this if it wasn't doing anything

**Comment by MHBauer** [comment]
Is there something I can do to help/test?

For context: I mainly use https://github.com/Mirantis/kubeadm-dind-cluster as it is very helpful to be able to get a `git HEAD` dev version of kubernetes up and running with a bunch of sane defaults.

**Comment by r2d4** [comment]
@minikube-bot retest this please

**Comment by r2d4** [comment]
This branch is a bit out of a date now - as I split most of the changes into smaller PRs (command runner, localkube bootstrapped). The last part of this is just splitting out the kubeadm changes to a new PR, which I should hopefully get to soon. I've just been a bit busy with other things.

**Comment by r2d4** [comment]
I've rebased and opened up #1903

**Pull request closed without merging**